### PR TITLE
GH-43608: [CI][Archery] Prefer `docker compose` over `docker-compose`

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

The `docker-compose` utility is legacy, while the Docker CLI client now includes a `compose` subcommand that attempts to be compatible with docker-compose YAML files.

### What changes are included in this PR?

1. Call `docker compose` by default when `archery docker` is invoked, not `docker-compose`
2. Add an option to switch back to `docker-compose`, in case of Docker CLI client is too old and doesn't support some options (such as `--file`).

### Are these changes tested?

Yes, on CI.

### Are there any user-facing changes?

No.
* GitHub Issue: #43608

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit *wheel*3.8* *cuda*

**Comment by pitrou** [comment]
@github-actions crossbow submit *wheel*cp38* *cuda*

**Comment by pitrou** [comment]
@github-actions crossbow submit *wheel*cp38* test-cuda* *jars* *skyhook* *arm*

**Comment by pitrou** [comment]
@github-actions crossbow submit *skyhook* wheel-windows*

**Comment by pitrou** [comment]
This draft PR disables `ARCHERY_USE_DOCKER_CLI` in a lot of places but we might actually want to keep wherever possible if it improves build caching (I have no idea). cc @kou @kszucs

File: dev/archery/archery/docker/cli.py [hunk 1]
```diff
@@ -47,18 +47,24 @@ def _execute(self, *args, **kwargs):
...
+@click.option('--using-legacy-docker-compose', default=False, is_flag=True,
+              envvar='ARCHERY_USE_LEGACY_DOCKER_COMPOSE',
+              help="Use legacy docker-compose utility instead of the built-in "
+                   "`docker compose` subcommand. This may be necessary if the "
+                   "Docker client is too old for some options.")
 @click.option('--using-docker-cli', default=False, is_flag=True,
               envvar='ARCHERY_USE_DOCKER_CLI',
               help="Use docker CLI directly for building instead of calling "
-                   "docker-compose. This may help to reuse cached layers.")
+                   "`docker compose`. This may help to reuse cached layers.")
```
Comment (kou) [inline dev/archery/archery/docker/cli.py:58]: Can we remove this option?

File: dev/tasks/tasks.yml [hunk 2]
```diff
@@ -1110,9 +1110,8 @@ tasks:
     template: docker-tests/github.linux.yml
     params:
       env:
-        ARCHERY_USE_DOCKER_CLI: 0
         UBUNTU: 20.04
-      flags: -e ARROW_SKYHOOK=ON
+      flags: -e ARROW_SKYHOOK=ON -e ARROW_AZURE=OFF -e ARROW_GCS=OFF -e ARROW_S3=OFF -e ARROW_GANDIVA=OFF
```
Comment (kou) [inline dev/tasks/tasks.yml:1114]: How about using `>-` and sort them?

```suggestion
      flags: >-
        -e ARROW_AZURE=OFF
        -e ARROW_GANDIVA=OFF
        -e ARROW_GCS=OFF
        -e ARROW_S3=OFF
        -e ARROW_SKYHOOK=ON
```

**Review by kou — COMMENTED** [review]
> This draft PR disables `ARCHERY_USE_DOCKER_CLI` in a lot of places

I'm OK with this for now. We can enables it by follow-up tasks.

> but we might actually want to keep wherever possible if it improves build caching (I have no idea).

I think that we need to use `docker compose` instead of `docker-compose` in all places eventually whether it improves build caching or not. Because `docker-compose` is obsolete.

File: dev/tasks/tasks.yml (hunk 2 above)
Comment (pitrou) [reply dev/tasks/tasks.yml:1114]: Nice suggestion!

File: dev/archery/archery/docker/cli.py (hunk 1 above)
Comment (pitrou) [reply dev/archery/archery/docker/cli.py:58]: Well, apparently it's necessary for Windows wheels:
https://github.com/ursacomputing/crossbow/actions/runs/10269082430/job/28413695620

**Comment by pitrou** [comment]
@github-actions crossbow submit *skyhook* *cuda* wheel-windows* *jars* *arm64*

**Comment by pitrou** [comment]
@github-actions crossbow submit test-cuda*

**Comment by pitrou** [comment]
@github-actions crossbow submit *jars* *wheel-windows*

File: dev/archery/archery/docker/cli.py (hunk 1 above)
Comment (pitrou) [reply dev/archery/archery/docker/cli.py:58]: Update: it's not if I ensure that buildkit is disabled.

**Comment by pitrou** [comment]
@kou How about we deprecate ARCHERY_USE_DOCKER_CLI in another issue/PR? This doesn't seem necessary right now.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g nightly-tests

**Comment by pitrou** [comment]
@github-actions crossbow submit -g nightly-packaging

**Comment by pitrou** [comment]
The remaining CI failures look unrelated to these changes.

**Review by kou — APPROVED** [review]
+1

> How about we deprecate ARCHERY_USE_DOCKER_CLI in another issue/PR?

I'm OK with this approach.

**Pull request merged**